### PR TITLE
Implement proper(?) Qt model semantics for dive filtering

### PR DESCRIPTION
**Pull request by bstoeger**
### Describe the pull request:

- [ ] Bug fix
- [ ] Functional change
- [ ] New feature
- [x] Code cleanup
- [ ] Build system change
- [ ] Documentation change
- [ ] Language translation

### Pull request long description:

This is the "more controversial" PR I was hinting at in the recent model-unification PR. The ultimate goal is to use the same models on desktop and mobile as far as possible.

What this does is sending the proper changed-events if the shown status of a dive changed. Thus, when the filter-criteria change it is not necessary to fully invalidate the filter. This might be controversial, because the code is rather complex. Nevertheless, to my understanding(!), this is how qt-models are intended to be implemented. At least Qt's own model-source is full of these complications.

Next up: moving down selection changes from desktop's DiveListView, so that can be reused on mobile as well.

### Changes made:

1) Send changed signals when show status changes.
2) Don't fully invalidate filter when filter-criteria change.
### Related issues:

Continuation of #2389.

### Release note:

No - should not be user visible.

### Mentions:

@dirkhh: please comment.

### CODE REVIEW

**Comment by bstoeger** [comment]
Selection related functions are now collected in their own file.

File: qt-models/divetripmodel.cpp [hunk 1]
```diff
@@ -693,7 +693,7 @@ dive *DiveTripModelTree::diveOrNull(const QModelIndex &index) const
 
 // Send data changed signals blockwise. Which entries are changed is collected
 // in the first parameter "changed".
-void DiveTripModelBase::sendShownChangedSignals(const std::vector<char> &changed, int parentIndex)
+void DiveTripModelBase::sendShownChangedSignals(const std::vector<char> &changed, quintptr parentIndex)
```
Comment (dirkhh) [inline qt-models/divetripmodel.cpp:710]: I've never noticed us using `quintptr` - what's the advantage over `int *` ?

**Review by dirkhh — COMMENTED** [review]
I haven't tested the code (I hope to do that soon) but a code review didn't find much that seemed all that controversial. I added some comments.

File: qt-models/divetripmodel.cpp (hunk 1 above)
Comment (bstoeger) [reply qt-models/divetripmodel.cpp:710]: quintptr is what Qt wants in its API:
https://doc.qt.io/qt-5/qabstractitemmodel.html#createIndex-1

This should have read quintptr from the start - I will rebase accordingly.

File: qt-models/divetripmodel.cpp [hunk 2]
```diff
@@ -691,6 +691,17 @@ dive *DiveTripModelTree::diveOrNull(const QModelIndex &index) const
 	return tripOrDive(index).dive;
 }
 
+static bool calculateFilterForTrip(const std::vector<dive *> &dives, const DiveFilter *filter)
+{
+	bool showTrip = false;
+	for (dive *d: dives) {
+		bool shown = filter->showDive(d);
+		filter_dive(d, shown);
+		showTrip |= shown;
```
[the remark this answers is not in the document]
Comment (bstoeger) [reply qt-models/divetripmodel.cpp:739]: Agreed.

File: qt-models/divetripmodel.cpp (hunk 1 above)
Comment (bstoeger) [reply qt-models/divetripmodel.cpp:710]: Note that, in contrast to what the name might imply, `quintptr` is an integer type, *not* a pointer. That's why we are using this here over the `void *` version. It's Qt's version of the C99 `uintptr_t` type.

Comment (dirkhh) [reply qt-models/divetripmodel.cpp:710]: mind boggles

**Comment by dirkhh** [comment]
my conclusion from the comments is that you're planning to rebase this and are waiting for me to do some testing, but overall we agree that this is going in the right direction, correct?

**Comment by bstoeger** [comment]
> my conclusion from the comments is that you're planning to rebase this and are waiting for me to do some testing, but overall we agree that this is going in the right direction, correct?

Yes. I added a bit of selection streamlining, please check it out as well. If you are fine, I will rebase the "mobile-list ng" branch and try to finally progress there.

**Comment by bstoeger** [comment]
Some more selection detangling (hopefully).

**Review by dirkhh — APPROVED** [review]
Done another round of reading the code and it seems fine to me.
I continue to be impressed with your ability to untangle this unholy mess.
I'll do some testing locally, but I think this is ready to be merged.

**Comment by dirkhh** [comment]
Ah, fun merge conflict.. :-)

**Comment by bstoeger** [comment]
Rebased on master, but didn't test the result.

As I said elsewhere: sometimes I have doubts whether this is untangling or running in circles. Let's see how it works out with mobile.

**Comment by dirkhh** [comment]
Thanks for the rebase :-)
I played with this some (not a lot) and it seems to work fine.
Is there a specific area that you are worried about? Or should we merge it and see what breaks 🙈

**Comment by bstoeger** [comment]
The part I worry about is the interaction of selection and filter. If you don't find anything broken there, I suggest to move on.

**Comment by bstoeger** [comment]
Sneaked in another trivial cleanup...

**Comment by bstoeger** [comment]
And two more code-hygiene things - sorry. 😛

**Review by dirkhh — APPROVED** [review]

**Comment by dirkhh** [comment]
this has a conflict because of the undo PR that I merged. I think I merged this correctly locally, but then again... I'd rather have you rebase it :-)
SORRY

**Comment by bstoeger** [comment]
Rebased on master. This fixes #2418, please commit.

**Comment by bstoeger** [comment]
Since you agreed with this PR, I'll merge so that we can progress.

**Pull request merged**